### PR TITLE
Fix get by tag

### LINKED ISSUE #31: Split out getByTag query into smaller functions or queries?
The current file is a little unwieldly and difficult to read - would be good to either separate them out into functions, or to just create completely new queries for each one.

### LINKED ISSUE #32: tag name not returned from getByTag
When no organisations match with the tag requested for, the response returns empty.

We ideally would return with just the tag_name so that we can display it in the header. 

Could be an enhancement.

### LINKED ISSUE #33: Formatter crashes if no data is returned from getByTag
Need to fix instances where getByTag query returns an empty array, and catch for when data[0] does not exist.

>> Could just check for the empty array and then make another request to just get tagName? But this would be a much less elegant solutio

### PR DESCRIPTION
**Pull request by mk4111**
Closes #31 
Closes #32 
Closes #33 

Fixes issue where app crashed when an empty array was being returned from the query, due to there being no orgs matching a certain tag.
Refactors query file so that the if/else is more modular and hopefully easier to read.
Adds tests for all 4 combinations of active(/inactive) and filter(/no filter) and tests for non-existent tag ids too.

### CODE REVIEW

**Comment by mk4111** [comment]
Hi @SimonLab and @JMurphyWeb - could you have a look at this on Monday please? Thanks!

File: tests/orgsGetByTag.test.js [hunk 1]
```diff
@@ -64,3 +144,52 @@ test('Get all the active organisations, associated with a specific tag', functio
...
+      url: '/orgsGetByTag?tags=' + tagId
+    }, function (res) {
+      var expected = {
+        filter_tag: 'Corporate',
+        orgs: [{
+          id: null,
+          name: null,
+          logo_url: null,
+          active: null
+        }]
```
Comment (jay-meister) [inline tests/orgsGetByTag.test.js]: @mk4111 i like this test, but it seems to make most sense that in this case, that orgs is an empty array. That way orgs.length will always equal the number of orgs returned.

File: tests/orgsGetByTag.test.js [hunk 2]
```diff
@@ -4,6 +4,40 @@ var test = require('tape');
...
+        'Charcoal',
+        'Coca Cola',
+        'Co-op Group',
+        'dwyl',
+        'EMF',
+        'Fanta',
+      ];
+      expected.forEach(function (org, index) {
+        var regexp = new RegExp(org, 'g');
+        t.ok(res.payload.match(regexp), 'org name ' + org + ' displays as expected');
```
Comment (jay-meister) [inline tests/orgsGetByTag.test.js]: Just wondering if there is any reason to use regex as opposed to indexOf?

**Comment by jay-meister** [comment]
I am sure it is better than it was, but I find the new query quite hard to follow still. It would really help in this case to see some query-string tests so that we can see exactly what is expected...

How much complexity is down to the `active = true` part?  We could easily filter results in the formatter or the handler if it helps simplify this file much?  

I think if there are tests showing each of the queries that this file produces, then we can make a better decision and more useful review.

Also we need to update to the readme with the function's new interface.

File: lib/index.js [hunk 3]
```diff
@@ -22,6 +22,9 @@ function register (server, options, next) {
         if (err) {
           return cb(err, null);
         }
+        if (rows.length === 0) { // i.e. no orgs -OR- filter --> wrong tagId or no orgs data
+          return cb(err, null);
```
Comment (jay-meister) [inline lib/index.js]: Im not sure what is best to return in this case. I thought no orgs would return the object still as we want to return the filter tag.

Doesn't it mean the tag Id was wrong? In which case what is best to return? is it an error?

Comment (jay-meister) [reply lib/index.js]: maybe @SimonLab can advise?

Comment (mk4111) [reply lib/index.js]: Hi @JMurphyWeb!
The reason why I have the catch `rows.length === 0` is for the case where someone might input an incorrect tagId. If no tag is found at the specified tagId, then `rows` returns an empty array. This then gets passed into `formatter.orgsGetByTag`, which breaks ( #33 ). We need to catch this here, (or we could move it into the formatting function), in order to return the data from the query, in the format that we expect it to be in.

File: tests/orgsGetByTag.test.js (hunk 2 above)
Comment (mk4111) [reply tests/orgsGetByTag.test.js]: Yeah I originally wanted to check that these org names would display, but also that they would only display once. I decided against it once I realised that that might not actually be the case but kept it as regex anyway.

File: tests/orgsGetByTag.test.js (hunk 1 above)
Comment (mk4111) [reply tests/orgsGetByTag.test.js]: I've actually changed this so it's now `orgs: null` if no orgs are returned. Am I right in guessing you'd still prefer an empty array? I guess it makes sense given the nature of the pg responses in general.

File: tests/orgsGetByTag.test.js (hunk 2 above)
Comment (jay-meister) [reply tests/orgsGetByTag.test.js]: Cool. Good to know.

File: lib/index.js (hunk 3 above)
Comment (jay-meister) [reply lib/index.js]: Yeah. I think it is good where it is, I am just wondering what we want to return in this case.

File: tests/orgsGetByTag.test.js (hunk 1 above)
Comment (jay-meister) [reply tests/orgsGetByTag.test.js]: I think empty array is more consistent with what we send back if there are no results. What are your thoughts?

File: lib/index.js (hunk 3 above)
Comment (jay-meister) [reply lib/index.js]: Null is probably fine, but I was wondering if an error would be suitable? I am not sure though, would be keen to hear your and Simons thoughts.

**Comment by mk4111** [comment]
Hi @JMurphyWeb! I've updated the code so that we're now throwing an error for invalid filtering requests. Could you have a quick review of the latest update and assign to Simon if happy? Thank you!

**Comment by jay-meister** [comment]
Ok just the readme and package version number.

You can assign Simon once that is done.

**Comment by mk4111** [comment]
@SimonLab could you please have a review of this one too please? Merci beaucoup!! :meat_on_bone:

**Review by SimonLab — APPROVED** [review]
Looks good, thanks @mk4111. Might be worth to keep in mind if passing active as a parameter simplify the code or if we should split the query and function.
